### DISCUSSION TITLE
VImage::new_from_file() causes Windows to fail to delete the file and raises ERROR_ACCESS_DENIED

### DISCUSSION THREAD

**Discussion by VivitionDeveloper** [libvips/libvips#4146] [Help wanted]
For some reason, even after destruct of the involved VImage object, the call to VImage::new_from_file() keeps something locked and (weird) Windows cannot delete the file. The Windows API function DeleteFile() always returns error 5 (ERROR_ACCESS_DENIED).

Is there a way to say to vips: close this file, release from cache or internal memory, and thus close all handles to the file system, so Windows can delete the file?

Pseudo code:
```
const char *szFile = "c:\\temp\\test.jpg";
{
  // variable in own scope, destructed before calling delete
  auto img = VImage::new_from_file(szFile);
}
const auto iErr = DeleteFile(szFile);

```

I found some other (older) issues regarding new_from_file, however, the fact that Linux is capable in deleting opened files is not a good reason to keep them open. Windows is unfortunately not capable in doing so. Also, the error does not indicate a lock, but indicates an access error. Please let me know how to solve this (or otherwise, pollute the file system).

Also, using sequential read does not work for us, as many of the libvips operations being used by us then cause an exception. I therefore must rely on the default open mode of libvips.

**Comment by jcupitt** [comment]
Hello again,

This is the libvips operation cache -- it's keeping the image open behind your back to make reuse quicker. If you disable the operation cache with `vips_cache_set_max(0);` it should close the file when `img` goes out of scope.

> Also, using sequential read does not work for us, as many of the libvips operations beinh used by us then cause an exception.

It's often (but not always) possible to restructure code to make sequential mode work. Could you give an example of something that fails? Things like 90 degree rotate won't ever work, of course :(

**Reply by VivitionDeveloper** [reply]
Thanks for reply and this provided solution. I will checkout this today, not sure whether the cache option must be set immediately after VIPS_INIT or that the cache can be disabled anywhere later on (possibly after other vips actions may have been performed). Can you please let me know?

And what about when I like to disable the cache only for one specific case? Can I turn on the cache to its default back? Or should I use the [VIPS_OPERATION_NOCACHE](https://www.libvips.org/API/current/VipsOperation.html#VipsOperationFlags) option when using new_from_file?

The example about the exception was provided some time ago in another [discussion](https://github.com/libvips/libvips/discussions/4036#discussioncomment-10094000).

**Reply by jcupitt** [reply]
You can set the cache size any time.

You can use the `revalidate` option to force `new_from_file()` to go back to the disc and not use the cache, would that work?

**Reply by VivitionDeveloper** [reply]
I do not know the default cache size, so when setting back, what value is then 'the right' value?

I'm not sure what you mean with use revalidate option, as I did not see such option at: https://www.libvips.org/API/current/VipsImage.html#vips-image-new-from-file
(pasting URLs is not a fun here, as it is reformatted upon paste, but hope the URL is right).

Anyhow: because I'm about the end of current sprint, it takes two weeks before I've time to check the option. But I will do so and let know the outcome, but when I know which options need to be set.

I was a little surprised to see the Britisch spelling of 'disc', while (here in NL) the more common 'disk' is being used, when a harddisk or SSD is involved. But nice to see. And also explains that some vips methods have the word 'disc' in it ;).

**Reply by jcupitt** [reply]
You can use `get_max()` to get the cache size.

You can see the `revalidate` here:

https://www.libvips.org/API/current/VipsForeignSave.html#VipsForeignSave.properties

Perhaps the general discussion at the start of that page should talk about it too.

**Reply by jcupitt** [reply]
Way, way back when libvips started I decided not to use US spelling because this thing was only for my use and I thought it didn't matter. These things are surprisingly hard to change after release 1.0!

**Reply by jcupitt** [reply]
Ah, or you can use eg.:

```
$ vips jpegload
load jpeg from file
usage:
   jpegload filename out [--option-name option-value ...]
where:
   filename     - Filename to load from, input gchararray
   out          - Output image, output VipsImage
optional arguments:
   shrink       - Shrink factor on load, input gint
			default: 1
			min: 1, max: 8
   autorotate   - Rotate image using exif orientation, input gboolean
			default: false
   unlimited    - Remove all denial of service limits, input gboolean
			default: false
   flags        - Flags for this file, output VipsForeignFlags
			default flags: 
			allowed flags: none, partial, bigendian, sequential, all
   memory       - Force open via memory, input gboolean
			default: false
   access       - Required access pattern for this file, input VipsAccess
			default enum: random
			allowed enums: random, sequential, sequential-unbuffered
   fail-on      - Error level to fail on, input VipsFailOn
			default enum: none
			allowed enums: none, truncated, error, warning
   revalidate   - Don't use a cached result for this operation, input gboolean
			default: false
```

So any of those options can be passed to `new_from_file` for JPEG images.

**Reply by VivitionDeveloper** [reply]
Many thanks again for your helpful reply.

**Reply by VivitionDeveloper** [reply]
It took some time before I was able to test, only disabling the cache (vips_cache_set_max(0)) worked + the vips image MUST reside in its own scope and must be destructed before the delete succeeds.

**Comment by VivitionDeveloper** [comment]
Some minor other thing: the documentation suffers from bad markup, i.e. HTML tags seem to be added as literal text. 

Examples:

 From: https://www.libvips.org/API/current/VipsImage.html#VipsImage.object-hierarchy:
![image](https://github.com/user-attachments/assets/df3c9710-6241-4409-a597-00330614e0b8)

From: https://www.libvips.org/API/current/VipsObject.html
![image](https://github.com/user-attachments/assets/e2cc4d8d-b950-4960-b949-23096e4614cf)

**Reply by jcupitt** [reply]
Yes, there are a few like that. The doc tool we use keeps "evolving" and changing its mind about what kind of markup it prefers, sigh.

**Reply by VivitionDeveloper** [reply]
There are others, even worse to read... But it is still understandable what is written there (example is from https://www.libvips.org/API/current/VipsImage.html#vips-image-new-from-file)
![image](https://github.com/user-attachments/assets/c747ad37-b9bd-4ae5-adf1-db7fa2990310)

**Reply by kleisauke** [reply]
Hmm, the 8.13 docs seems to be OK:
https://www.libvips.org/API/8.13/VipsImage.html#VipsImage.object-hierarchy
https://www.libvips.org/API/8.13/VipsObject.html#VipsObject.description
https://www.libvips.org/API/8.13/libvips-arithmetic.html#vips-divide

Both were generated by GTK-Doc v1.33.1, so this could be a some sort of configuration issue in the Meson build. I'll investigate.

**Reply by jcupitt** [reply]
We should probably switch to markdown in all gtkdoc comments. HTML Tables are very annoying to enter by hand.

**Reply by kleisauke** [reply]
PR #4149, more specifically commit fdd433663423b60a1ccbb4c4a896c93636f53d33, will fix this. See e.g. the diff in PR #4150.

**Reply by VivitionDeveloper** [reply]
Great, thank you.
